### ISSUE TITLE
CRF library does not work as expected

### ISSUE THREAD

**Issue by vijay120** [kmkurn/pytorch-crf#62]
I am using pytorch-crf to resolve tagging issues in NER, where the `I-entity` tag gets predicted after `O` (eg. `O, I-person_name`) or before `B-entity` tags (eg. `I-person_name, B-person_name`). Even after using pytorch-cr, I still am seeing decoded tags with the same errors as before.

I inspected the `transitions`, `start_transitions` and `end_transition` attributes of the CRF library (https://github.com/kmkurn/pytorch-crf/blob/master/torchcrf/__init__.py), and found that the library does not update any of these variables. They seem to be randomly initialized and then called for scoring, but these state transition variables are never updated.

Can you please provide a better understanding on why this is case, and probably provide some input on why these tag errors keep happening in-spite of using the CRF layer.

Thanks and Sincerely,
Vijay Ramakrishnan

**Comment by kmkurn** [comment]
Hi, I tested it and the variables' gradients were computed properly, so they should be updated. Can you provide a simple example where the they are not?

**Comment by LorenzoPozzi97** [comment]
Hi @vijay120, how did you solve the problem? I also have this issue and the start, end and transition score do not update.
`for epoch_i in range(epochs):  
    model.train()
    for i, v in train_data.items():
      model.zero_grad()
      optimizer.zero_grad()

      # Step 2. Run the forward pass.
      train_input_ids = train_data[i]['input_ids'].to(device)                    # (batch, max_seq_length)
      train_input_mask = train_data[i]['attention_mask'].to(device)              # (batch, max_seq_length)
      train_input_type = train_data[i]['token_type_ids'].to(device)              # (batch, max_seq_length)
      token_train_label = train_labels['Tags'][i].to(device)                     # (batch, max_seq_length)
      
      scores = model.forwardCRF(train_input_ids.long(), 
                                    attention_mask = train_input_mask.long(), 
                                    token_type_ids = train_input_type.long())     # (batch, max_seq_length, 2)
          
      log_lik = -model.crf(scores.long(), token_train_label.long(), train_input_mask.bool(), reduction='mean')
      prediction = model.crf.decode(scores.long(), train_input_mask.bool())
      
      log_lik.backward()
      optimizer.step()`

@kmkurn Could you pls help me with this issue? Am I missing something fundamental?

**Comment by LorenzoPozzi97** [comment]
This is the model @kmkurn . It is a stack of BERT-LSTM-CRF:
`class IndexerBert(nn.Module):
    def __init__(self, config, doCRF = 'off'):
      super(IndexerBert, self).__init__()
      self.num_labels = config.num_labels
      self.LSTM_hidden_dim = 250
      self.bert = BertModel.from_pretrained("bert-base-uncased")
      self.dropout = nn.Dropout(p=0.5)
      self.lstm = nn.LSTM(input_size = config.hidden_size,  
                              hidden_size = self.LSTM_hidden_dim, 
                              batch_first=True, 
                              bidirectional=True,
                              dropout=0.05)
     
      self.crf = CRF(self.num_labels, batch_first=True)
      self.classifier2 = nn.Linear(self.LSTM_hidden_dim*2, 2)
      
     
     
      
      # Freeze BERT parameters
      for name, param in list(self.bert.named_parameters())[:-50]: 
        param.requires_grad = False

    def forwardCRF(self, input_ids, attention_mask = None, token_type_ids = None):
      output, pooled_output = self.bert(input_ids, 
                                        attention_mask, 
                                        token_type_ids,
                                        return_dict=False)
      
      dropout_output = self.dropout(output)
      input, (h_0, c_0) = self.lstm(dropout_output)                              # (batch, seq_len, 500)
      
      linear_output = self.classifier2(input)                                    # (batch, seq_len, 2)
      
      return linear_output       `
